chips/earlgrey/src/epmp.rs

```diff
@@ -384,6 +438,116 @@ impl<const HANDOVER_CONFIG_CHECK: bool> EarlGreyEPMP<{ HANDOVER_CONFIG_CHECK },
+
+        // Set the appropriate `pmpcfg0` register value:
+        //
+        // 0x80 = 0b10000000, for start the address of the kernel .text TOR
+        //        entry as well as entries 0 and 1.
+        //        setting L(7) = 1, A(4-3) = OFF, X(2) = 0, W(1) = 0, R(0) = 0
+        //
+        // 0x8d = 0b10001101, for kernel .text TOR region
+        //        setting L(7) = 1, A(4-3) = TOR,   X(2) = 1, W(1) = 0, R(0) = 1
+        //
+        // Note that we try to lock entries 0 and 1 into OFF mode. If the
+        // ROM_EXT set these up and locked them, this will do nothing, otherwise
+        // it will permanently disable these entries (preventing them from being
+        // misused later).
+        csr::CSR.pmpcfg0.set(0x8d_80_80_80);
+
+        // --> Continue with the "low-priority" ("accessibility") section of the
+        // ePMP configuration:
+
+        // Configure a Read-Only NAPOT region for the entire flash (spanning
+        // kernel & apps, but overlayed by the R/X kernel text TOR section)
+        csr::CSR.pmpaddr12.set(flash.0.napot_addr());
+
+        // Configure a Read-Write NAPOT region for MMIO.
+        csr::CSR.pmpaddr14.set(mmio.0.napot_addr());
+
+        // Configure a Read-Write NAPOT region for the entire RAM (spanning
+        // kernel & apps)
+        csr::CSR.pmpaddr15.set(ram.0.napot_addr());
+
+        // With the FLASH, RAM and MMIO configured in separate regions, we can
+        // activate this new configuration, and further adjust the permissions
+        // of the (currently all-capable) last PMP entry `pmpaddr15` to be R/W,
+        // as required for MMIO:
+        //
+        // 0x99 = 0b10011001, for FLASH NAPOT region
+        //        setting L(7) = 1, A(4-3) = NAPOT, X(2) = 0, W(1) = 0, R(0) = 1
+        //
+        // 0x80 = 0b10000000, for the unused region
+        //        setting L(7) = 1, A(4-3) = OFF, X(2) = 0, W(1) = 0, R(0) = 0
+        //
+        // 0x9B = 0b10011011, for RAM & MMIO NAPOT regions
+        //        setting L(7) = 1, A(4-3) = NAPOT, X(2) = 0, W(1) = 1, R(0) = 1
+        csr::CSR.pmpcfg3.set(0x9B_9B_80_99);
+
+        // Ensure that the other pmpcfgX CSRs are cleared:
+        csr::CSR.pmpcfg1.set(0x00000000);
+        csr::CSR.pmpcfg2.set(0x00000000);
+
+        // ---------- PMP machine CSRs configured, lock down the system
+
+        // Finally, unset the rule-lock bypass (RLB) bit. If we don't have a
+        // debug memory region provided, further set machine-mode lockdown (we
+        // can't enable MML and also have a R/W/X region). We also set MMWP for
+        // good measure, but that shouldn't make a difference -- it can't be
+        // cleared anyways as it is a sticky bit.
+        //
+        // Unsetting RLB with at least one locked region will mean that we can't
+        // set it again, thus actually enforcing the region lock bits.
+        //
+        // Set RLB(2) = 0, MMWP(1) = 1, MML(0) = 1
```

Comment:
From what I understand, the new `ROM_EXT` will already have this exact(?) `mseccfg` CSR value set for us. If that is the case, I'm by no means opposed to keeping this write in to make this explicit, but we might want to update the comment documenting this expectation.

Reply:
The `ROM_EXT` I'm working with actually sets `0x2` (MMWP, but not MML or RLB). I've updated the documentation to just say we're enabling MML.